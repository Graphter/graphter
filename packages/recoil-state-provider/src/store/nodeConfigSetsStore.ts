import { atom, RecoilState } from "recoil";
import { NodeConfig, PathSegment } from "@graphter/core";
import { pathToKey } from "@graphter/renderer-react";

export interface NodeConfigSets {
  activeConfigsKey: string,
  configSets: Map<string, Array<NodeConfig>>
}

const pathNodeConfigMap = new Map<string, RecoilState<NodeConfigSets>>()

const getPathKey = (path: Array<PathSegment>) => `node-configs-store-a12b32f8-ecb8-4522-bc9a-f1ab8a59caae-${pathToKey(path)}`

export const get = (
  path: Array<PathSegment>
): RecoilState<NodeConfigSets> => {
  checkPathArg(path)
  const key = getPathKey(path)
  if(!pathNodeConfigMap.has(key)) throw new Error(`Couldn't find configs for node at '${path.join('/')}'`)
  const state = pathNodeConfigMap.get(key)
  if(typeof state === 'undefined') throw new Error('Should not happen')
  return state
}

export const set = (
  path: Array<PathSegment>,
  configSets: NodeConfigSets
) => {
  checkPathArg(path)
  const pathKey = getPathKey(path)
  if(pathNodeConfigMap.has(pathKey)) throw new Error(`A value is already set for '${path.join('/')}'`)
  pathNodeConfigMap.set(pathKey, atom({
    key: pathKey,
    default: configSets
  }))
}

export const has = (path: Array<string | number>): boolean => {
  checkPathArg(path)
  return pathNodeConfigMap.has(getPathKey(path))
}

export const remove = (path: Array<PathSegment>) => {
  checkPathArg(path)
  pathNodeConfigMap.delete(getPathKey(path))
}


function checkPathArg(path: Array<PathSegment>){
  if(!path || !path.length) throw new Error('Path of at least one segment is required')
}

export const nodeConfigSetsStore = {
  get,
  set,
  has,
  remove
}