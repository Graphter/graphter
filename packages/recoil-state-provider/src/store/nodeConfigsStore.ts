import { atom, RecoilState } from "recoil";
import { NodeConfig, PathSegment } from "@graphter/core";
import { pathToKey } from "@graphter/renderer-react";

const pathNodeConfigMap = new Map<string, RecoilState<Array<NodeConfig>>>()

const getKey = (path: Array<PathSegment>) => `node-configs-store-a12b32f8-ecb8-4522-bc9a-f1ab8a59caae-${pathToKey(path)}`

export const get = (
  path: Array<PathSegment>
): RecoilState<Array<NodeConfig>> | null => {
  checkPathArg(path)
  const key = getKey(path)
  if(!pathNodeConfigMap.has(key)) return null
  const state = pathNodeConfigMap.get(key)
  if(typeof state === 'undefined') throw new Error('Should not happen')
  return state
}

export const set = (
  path: Array<PathSegment>,
  configs: Array<NodeConfig>
) => {
  console.log(`Setting configs for path '${path.join('/')}' to ${JSON.stringify(configs)}`)
  checkPathArg(path)

  const pathKey = getKey(path)

  if(pathNodeConfigMap.has(pathKey)) throw new Error(`A value is already set for '${path.join('/')}'`)

  pathNodeConfigMap.set(pathKey, atom({
    key: pathKey,
    default: configs
  }))
}

export const has = (path: Array<string | number>): boolean => {
  checkPathArg(path)
  return pathNodeConfigMap.has(getKey(path))
}

export const remove = (path: Array<PathSegment>) => {
  checkPathArg(path)
  pathNodeConfigMap.delete(getKey(path))
}


function checkPathArg(path: Array<PathSegment>){
  if(!path || !path.length) throw new Error('Path of at least one segment is required')
}

export const nodeConfigsStore = {
  get,
  set,
  has,
  remove
}