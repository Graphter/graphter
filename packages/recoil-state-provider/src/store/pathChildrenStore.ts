import { atom, RecoilState } from "recoil";
import { PathSegment } from "@graphter/core";
import { pathToKey } from "@graphter/renderer-react";

const childPathsMap = new Map<string, RecoilState<Array<any>>>()

const getKey = (path: Array<PathSegment>) => `path-children-store-a4d28666-5de8-4825-bcda-b1b6d116abe8-${pathToKey(path)}`

export const get = (
  path: Array<PathSegment>
): RecoilState<Array<any>> => {
  const key = getKey(path)
  if(!childPathsMap.has(key)) throw new Error(`Couldn't find path children state at '${path.join('/')}'`)
  const state = childPathsMap.get(key)
  if(typeof state === 'undefined') throw new Error('Should not happen')
  return state
}

export const set = (
  parentPath: Array<PathSegment>,
  childPaths: Array<Array<PathSegment>>
) => {

  const pathKey = getKey(parentPath)

  if(childPathsMap.has(pathKey)) throw new Error(`Tree path state already exists for '${parentPath.join('/')}'`)

  childPathsMap.set(pathKey, atom({
    key: pathKey,
    default: childPaths
  }))
}

export const has = (parentPath: Array<string | number>): boolean => {
  return childPathsMap.has(getKey(parentPath))
}

export const remove = (parentPath: Array<PathSegment>) => {
  childPathsMap.delete(getKey(parentPath))
}

export const pathChildrenStore = {
  get,
  set,
  has,
  remove
}