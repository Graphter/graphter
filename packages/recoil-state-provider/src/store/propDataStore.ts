/***
 * TODO: This module makes some pretty nasty data mutations. Fix that.
 */

import { atom, RecoilState } from "recoil";
import { PathSegment } from "@graphter/core";
import { nanoid } from "nanoid";

interface PropDataStateMeta {
  state: RecoilState<any>,
  deleted: boolean,
  committed: boolean,
  default: any
}

interface Node {
  id: string | null,
  meta: PropDataStateMeta | null,
  children: Array<Node> | null
}

const propStateNodeTree: Array<Node> = []

export const init = (
  path: Array<PathSegment>,
  data: any
): void => {
  const init = (path: Array<PathSegment>, data: any) => {
    if(Array.isArray(data)) {
      if(!has(path)) set(path, true, data.map(() => nanoid()))
      data.forEach((item, i) => init([ ...path, i ], item))
    } else if(typeof data === 'object') {
      if(!has(path)) set(path, true, data)
      Object.entries(data).forEach(([key, value]) => {
        init([ ...path, key ], value)
      })
    } else {
      if(!has(path)) set(path, true, data)
    }
  }
  init(path, data)
}

export const get = (
  path: Array<PathSegment>
): RecoilState<any> => {
  checkPathArg(path)
  const pathNode = getPathNode(path)
  if(!pathNode.meta) throw new Error(`Couldn't find state at ${path.join('/')}`)
  return pathNode.meta.state
}

export const getAll = (
  path: Array<PathSegment>
): Array<RecoilState<any>> => {
  checkPathArg(path)
  const pathNode = getPathNode(path)
  if(!pathNode.children) throw new Error(`Couldn't find child state at ${path.join('/')}`)
  return Array.from(pathNode.children.values())
    .filter(node => node.meta && node.meta.committed)
    .map(node => node.meta?.state as RecoilState<any>)
}

export const set = (
  path: Array<PathSegment>,
  committed: boolean,
  originalValue?: any
) => {
  checkPathArg(path)
  console.info(`Setting ${path.join('/')} = ${originalValue} [committed=${committed}]`)
  const node = getPathNode(path)
  node.meta = {
    state: atom({
      key: nanoid(),
      default: originalValue
    }),
    deleted: false,
    committed,
    default: originalValue
  }
}

/***
 * The "committed" flag in Graphter is used to demarcate draft tree additions. For example
 * the "add item" ui for a list before the "Add [+]" button is pressed.
 *
 * In future we'll likely have to integrate this function with node modification history or
 * versioning to allow us to track tree modifications as well as additions. For example a
 * renderer may want to allow changes and then provide a Save/Cancel button to commit or
 * discard the tree changes
 * @param path
 */
export const commitItem = (
  path: Array<PathSegment>
) => {
  checkPathArg(path)
  const node = getPathNode(path)
  if(!node.meta) throw new Error(`Node at ${path.join('/')} is missing meta. Should never happen`)
  node.meta.committed = true
}

export const has = (path: Array<string | number>): boolean => {
  checkPathArg(path)
  let node: Node | undefined = {
    id: null,
    meta: null,
    children: propStateNodeTree
  }
  for(let segment of path){
    if(typeof segment === 'string'){
      node = node.children?.find(childNode => childNode.id === segment)
    } else {
      node = node.children ? node.children[segment] : undefined
    }

    if(!node) return false
  }
  return true
}

export const remove = (path: Array<PathSegment>) => {
  checkPathArg(path)
  const clonedPath = [...path]
  const removeSegment = clonedPath.splice(-1, 1)[0]

  const parentNode = getPathNode(clonedPath)
  if(!parentNode) throw new Error(`Missing ancestor trying to set ${path.join('/')}`)
  if(!parentNode.children) throw new Error(`No children at ${path.join('/')} to remove`)
  const indexToRemove = typeof removeSegment === 'number' ?
    removeSegment :
    parentNode.children
      .findIndex(childNode => childNode.id !== removeSegment)
  parentNode.children.splice(indexToRemove, 1)
}

const getPathNode = (path: Array<PathSegment>) => {
  return path.reduce<Node>((a, c) => {
    if(!a.children) a.children = []
    let node = typeof c === 'string' ?
      a.children.find(node => node.id === c) :
      a.children[c]

    if(!node){
      node = {
        id: typeof c === 'string' ? c : null,
        meta: null,
        children: null
      }
      typeof c === 'string' ?
        a.children.push(node) :
        a.children[c] = node
    }
    return node
  }, { id: null, meta: null, children: propStateNodeTree })
}

function checkPathArg(path: Array<PathSegment>){
  if(!path || !path.length) throw new Error('Path of at least one segment is required')
}

export interface PropDataStore {
  init: (path: Array<PathSegment>, data: any) => void
  get: (path: Array<PathSegment>) => RecoilState<any>
  getAll: (path: Array<PathSegment>) => Array<RecoilState<any>>
  set: (
    path: Array<PathSegment>,
    committed: boolean,
    originalValue?: any
  ) => void
  commitItem: (path: Array<PathSegment>) => void
  has: (path: Array<string | number>) => boolean
  remove: (path: Array<PathSegment>) => void
}

export const propDataStore: PropDataStore = {
  init,
  get,
  getAll,
  set,
  commitItem,
  has,
  remove
}

export default PropDataStore