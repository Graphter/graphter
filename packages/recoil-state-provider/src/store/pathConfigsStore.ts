import { selector, RecoilValueReadOnly } from "recoil";
import { NodeConfig, PathSegment } from "@graphter/core";
import { pathToKey } from "@graphter/renderer-react";
import { getPathPaths } from "../utils/getPathPaths";
import { nodeConfigSetsStore } from "./nodeConfigSetsStore";

const pathConfigsMap = new Map<string, RecoilValueReadOnly<Array<Array<NodeConfig>>>>()

export const get = (
  path: Array<PathSegment>
): RecoilValueReadOnly<Array<Array<NodeConfig>>> => {
  const pathKey = `path-config-store-${pathToKey(path)}`
  let pathConfigsSelector = pathConfigsMap.get(pathKey)
  if(pathConfigsSelector) return pathConfigsSelector
  pathConfigsSelector = selector({
    key: pathKey,
    get: ({ get }) => {
      const pathPaths = getPathPaths(path)
      return pathPaths.reduce<Array<Array<NodeConfig>>>((a, c) => {
        const nodeConfigSetsState = nodeConfigSetsStore.get(c)
        const nodeConfigSets = get(nodeConfigSetsState)
        const configs = nodeConfigSets.configSets.get(nodeConfigSets.activeConfigsKey)
        if (configs) a.push(configs)
        return a
      }, [])
    }
  })
  pathConfigsMap.set(pathKey, pathConfigsSelector)
  return pathConfigsSelector
}

export const pathConfigsStore = {
  get,
}