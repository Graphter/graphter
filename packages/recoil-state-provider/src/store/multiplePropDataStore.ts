import { RecoilValueReadOnly, selector } from "recoil";
import { PathSegment } from "@graphter/core";
import { propDataStore } from "./propDataStore";

const multiplePropDataMap: Map<string, RecoilValueReadOnly<any>> = new Map()
const keySalt = '5042b962-fd6c-470c-bbae-747380814365'
const get = (paths: Array<Array<PathSegment>>): RecoilValueReadOnly<Array<{ path: Array<PathSegment>, data: any }>> => {
  const key = paths.map(path => path.join(keySalt)).join('-')
  if(!multiplePropDataMap.has(key)){
    multiplePropDataMap.set(key, selector<Array<{ path: Array<PathSegment>, data: any }>>({
      key: key,
      get: ({ get }) => {
        return paths.map(path => {
          const state = propDataStore.get(path)
          if(!state) throw new Error(`Missing state for ${path.join('/')}'`)
          const pathMeta = get(state)
          return {
            path,
            data: pathMeta
          }
        })
      }
    }))
  }
  const state = multiplePropDataMap.get(key)
  if(!state) throw new Error(`Error gathering state for paths: ${paths.map(path => path.join('/')).join(' - ')}`)
  return state
}

const multiplePropDataStore = {
  get
}

export default multiplePropDataStore