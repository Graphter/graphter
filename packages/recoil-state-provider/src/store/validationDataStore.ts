import { NodeConfig, PathSegment, ValidationResult } from "@graphter/core";
import { atom, RecoilState, RecoilValueReadOnly, selector } from "recoil";
import { NodeValidationData } from "@graphter/renderer-react";
import { propDataStore } from "./propDataStore";

const validationStateMap: Map<string, RecoilState<NodeValidationData>> = new Map()
const pathSegmentKeySalt = '[be3cc399-9225-4cbc-9d23-6c0c69ddca4a]'
const aggregateValidationsStateMap: Map<string, RecoilValueReadOnly<Array<NodeValidationData>>> = new Map()
const pathKeySalt = '[c42e99e6-d234-4d4a-828f-f696b11c9683]'

const generateValidationKey = (path: Array<PathSegment>) => `validation-${path.join(pathSegmentKeySalt)}`

export const get = (path: Array<PathSegment>): RecoilState<NodeValidationData> => {
  const state = validationStateMap.get(generateValidationKey(path))
  if(!state) throw new Error(`No validation state for '${path.join('/')}' found in the store`)
  return state
}

export const getAll = (paths: Array<Array<PathSegment>>): RecoilValueReadOnly<Array<NodeValidationData>> => {
  const key = `aggregate-validations-${paths.map(path => path.join(pathSegmentKeySalt)).join(pathKeySalt)}`
  if(aggregateValidationsStateMap.has(key)){
    const aggregateValidations = aggregateValidationsStateMap.get(key)
    if(!aggregateValidations) throw new Error('Could not find aggregate validation. Should never happen')
    return aggregateValidations
  }

  const aggregateValidations = selector<Array<NodeValidationData>>({
    key,
    get: ({ get: selectorGet }) => {
      return paths.flatMap(path => {
        if(!has(path)) {
          if(!propDataStore.has(path)) return []
          const propState = propDataStore.get(path)
          const propData = selectorGet(propState)
          set(path, propData, [])
        }
        const validationDataState = get(path)
        const validationData = selectorGet(validationDataState)
        return [{
          path,
          results: validationData.results
        }]
      })
    }
  })
  aggregateValidationsStateMap.set(key, aggregateValidations)

  return aggregateValidations
}

export const has = (path: Array<PathSegment>): boolean => {
  return validationStateMap.has(generateValidationKey(path))
}

export const set = (
  path: Array<PathSegment>,
  value: any,
  results: Array<ValidationResult>
) => {
  if(!path || !path.length) throw new Error('path argument is required')
  const key = generateValidationKey(path)
  const validationData: NodeValidationData = {
    path,
    value,
    results
  }
  validationStateMap.set(key, atom({
    key: key,
    default: validationData
  }))
}

export interface ValidationDataStore{
  get: (path: Array<PathSegment>) => RecoilState<NodeValidationData>
  getAll: (paths: Array<Array<PathSegment>>) => RecoilValueReadOnly<Array<NodeValidationData>>
  has: (path: Array<PathSegment>) => boolean
  set: (
    path: Array<PathSegment>,
    value: any,
    results: Array<ValidationResult>
  ) => void
}

const validationDataStore: ValidationDataStore = {
  get,
  getAll,
  has,
  set
}

export default validationDataStore