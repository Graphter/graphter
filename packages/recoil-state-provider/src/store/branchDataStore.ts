import { RecoilValueReadOnly, selector } from "recoil";
import { PathSegment } from "@graphter/core";
import { nodeConfigSetsStore } from "./nodeConfigSetsStore";
import { pathChildrenStore } from "./pathChildrenStore";
import { nodeRendererStore } from "@graphter/renderer-react";
import { rendererInternalDataStore } from "./rendererInternalDataStore";
import { pathToKey } from "@graphter/renderer-react";
import { pathConfigsStore } from "./pathConfigsStore";
import { getExactPathConfigs } from "../utils/getExactPathConfigs";
import { pathConfigsToString } from "../utils/pathConfigsToString";

const treeDataMap: Map<string, RecoilValueReadOnly<any>> = new Map()

const get = <T>(startingPath: Array<PathSegment>, depth?: number) => {
  if (!startingPath) throw new Error('Path is required to get descendent data')
  const key = pathToKey(startingPath)
  let treeDataSelector = treeDataMap.get(key)
  if (treeDataSelector) return treeDataSelector

  treeDataSelector = selector<T>({
    key: key,
    get: ({get}) => {
      function getNodeData(path: Array<PathSegment>): any {
        const nodeConfigSetsState = nodeConfigSetsStore.get(path)
        if (!nodeConfigSetsState) throw new Error(`Missing path config state at path ${path.join('/')}`)
        const nodeConfigSets = get(nodeConfigSetsState)
        const nodeConfigs = nodeConfigSets.configSets.get(nodeConfigSets.activeConfigsKey)
        if(!nodeConfigs) throw new Error(`Can't find active configs at '${path.join('/')}'`)

        function getChildData(){
          if(!pathChildrenStore.has(path)) return null
          const childPathsState = pathChildrenStore.get(path)
          if (!childPathsState) return null
          const childPaths = get<Array<Array<PathSegment>>>(childPathsState)
          return childPaths.length ?
            childPaths.map(childPath => getNodeData(childPath)) :
            null
        }

        const childData = getChildData()
        const pathConfigsState = pathConfigsStore.get(path)
        const pathConfigs = get(pathConfigsState)
        // transform from the bottom most path node -> up
        const externalNodeData = [ ...nodeConfigs ].reverse().reduce<any>((a, c) => {
          const exactPathConfigs = getExactPathConfigs(pathConfigs, c)
          const internalDataState = rendererInternalDataStore.get(path, exactPathConfigs)
          if (!internalDataState){
            throw new Error(`Missing internal data for ${
              pathConfigsToString(exactPathConfigs)
            } at path '${path.join('/')}' node ${c.id}`)
          }
          const internalData = get(internalDataState)
          const rendererReg = nodeRendererStore.get(c.type)
          const rendererInternalData = typeof internalData !== 'undefined' ? internalData : a
          if (!childData?.length || !rendererReg.mergeChildData) return rendererInternalData
          else return rendererReg.mergeChildData(c, path, rendererInternalData, getNodeData, childData)
        }, undefined)

        return externalNodeData
      }

      return getNodeData(startingPath)
    }
  });
  treeDataMap.set(key, treeDataSelector)

  return treeDataSelector
}

const has = (startingPath: Array<PathSegment>) => {
  if (!startingPath) throw new Error('Path is required to get descendent data')
  const key = pathToKey(startingPath)
  return treeDataMap.has(key)
}

export const branchDataStore = {
  get,
  has
}

export default branchDataStore