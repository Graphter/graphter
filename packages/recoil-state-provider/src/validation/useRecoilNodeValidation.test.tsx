import { NodeConfig, NodeValidatorRegistration, ValidationExecutionStage } from "@graphter/core";
import React from "react";
import { render } from "@testing-library/react";
import { RecoilRoot } from 'recoil'
import { NodeValidationHook } from "@graphter/renderer-react";
import { ValidationDataStore } from "../store/validationDataStore";
import flushPromises from "../__test-utils__/flushPromises";

describe('useRecoilNodeValidation()', () => {
  let useRecoilNodeValidation: NodeValidationHook,
    validatorRegistration: NodeValidatorRegistration,
    validatorMock: jest.Mock<any>,
    validationDataStore: ValidationDataStore,
    propDataStore: any

  function ConsumerComponent({ cb, config }: { cb?: (data: any) => void, config: NodeConfig }) {
    const validationData = useRecoilNodeValidation(config, [ 'page' ], [ validatorRegistration ])
    if(cb) cb(validationData)
    return null
  }

  beforeEach(() => {
    jest.resetAllMocks()
    jest.isolateModules(() => {
      useRecoilNodeValidation = require('./useRecoilNodeValidation').useRecoilNodeValidation
      validationDataStore = require('../store/validationDataStore')
      propDataStore = require('../store/propDataStore')
    })
    validatorMock = jest.fn()
    validatorRegistration = {
      type: 'test-validation',
      validatorSetup: () => {
        return validatorMock
      }
    }
  })

  it('should return validation data for a node', async () => {
    propDataStore.set(['page'], true, 'the-page-data')
    const config = {
      id: 'page',
      type: 'object',
      validation: [
        {
          type: 'test-validation',
          executeOn: ValidationExecutionStage.CHANGE,
          options: { some: 'validation-constraints' }
        },
        {
          type: 'test-validation',
          executeOn: ValidationExecutionStage.CHANGE,
          options: { someOther: 'validation-constraints' }
        }
      ]
    }
    validationDataStore.set(['page'], 'the-page-data', [])
    validatorMock.mockResolvedValueOnce({
      valid: false,
      errorMessage: 'Some error message'
    })
    validatorMock.mockResolvedValueOnce({
      valid: false,
      errorMessage: 'Some other error message'
    })
    const cbMock = jest.fn()
    render(
      <RecoilRoot>
        <ConsumerComponent cb={cbMock} config={config} />
      </RecoilRoot>
    )
    await flushPromises()
    expect(cbMock).toHaveBeenCalledTimes(3)
    const result = cbMock.mock.calls[2][0]
    expect(result.results.length).toBe(2)
    expect(result.results[0].valid).toBe(false)
    expect(result.results[0].errorMessage).toBe('Some error message')
    expect(result.results[1].valid).toBe(false)
    expect(result.results[1].errorMessage).toBe('Some other error message')
  })
  it('should skip any validator other than those configured to run onChange', async () => {
    propDataStore.set(['page'], true, 'the-page-data')
    const config = {
      id: 'page',
      type: 'object',
      validation: [
        {
          type: 'test-validation',
          executeOn: ValidationExecutionStage.CHANGE,
          options: { some: 'validation-constraints' }
        },
        {
          type: 'test-validation',
          executeOn: ValidationExecutionStage.CLIENT_CREATE,
          options: { someOther: 'validation-constraints' }
        }
      ]
    }
    validationDataStore.set(['page'], 'the-page-data', [])
    validatorMock.mockResolvedValueOnce({
      valid: false,
      errorMessage: 'Some error message'
    })
    const cbMock = jest.fn()
    render(
      <RecoilRoot>
        <ConsumerComponent cb={cbMock} config={config} />
      </RecoilRoot>
    )
    await flushPromises()
    expect(cbMock).toHaveBeenCalledTimes(3)
    const result = cbMock.mock.calls[2][0]
    expect(result.results.length).toBe(1)
    expect(result.results[0].valid).toBe(false)
    expect(result.results[0].errorMessage).toBe('Some error message')
  })
  it('should return empty validation data if no onChange validators are found', async () => {
    propDataStore.set(['page'], true, 'the-page-data')
    const config = {
      id: 'page',
      type: 'object',
      validation: [
        {
          type: 'test-validation',
          executeOn: ValidationExecutionStage.SERVER_UPDATE,
          options: { some: 'validation-constraints' }
        },
        {
          type: 'test-validation',
          executeOn: ValidationExecutionStage.CLIENT_CREATE,
          options: { someOther: 'validation-constraints' }
        }
      ]
    }
    validationDataStore.set(['page'], 'the-page-data', [])
    validatorMock.mockResolvedValueOnce({
      valid: false,
      errorMessage: 'Some error message'
    })
    const cbMock = jest.fn()
    render(
      <RecoilRoot>
        <ConsumerComponent cb={cbMock} config={config} />
      </RecoilRoot>
    )
    await flushPromises()
    expect(cbMock).toHaveBeenCalledTimes(2)
    const result = cbMock.mock.calls[1][0]
    expect(result.results.length).toBe(0)
  })
  it('should throw an error if the node has no state', async () => {
    validationDataStore.set(['page'], 'the-page-data', [])
    validatorMock.mockResolvedValueOnce({
      valid: false,
      errorMessage: 'Some error message'
    })
    const cbMock = jest.fn()
    expect(() => render(
      <RecoilRoot>
        <ConsumerComponent cb={cbMock} config={{id: 'some-id', type: 'some-type'}} />
      </RecoilRoot>
    )).toThrowErrorMatchingSnapshot()
  })
  it('should set initial validationDataStore data if none exists yet', async () => {
    propDataStore.set(['page'], true, 'the-page-data')
    const config = {
      id: 'page',
      type: 'object',
      validation: [
        {
          type: 'test-validation',
          executeOn: ValidationExecutionStage.CHANGE,
          options: { some: 'validation-constraints' }
        }
      ]
    }
    validationDataStore.set(['page'], 'the-page-data', [])
    validatorMock.mockResolvedValueOnce({
      valid: false,
      errorMessage: 'Some error message'
    })
    const cbMock = jest.fn()
    render(
      <RecoilRoot>
        <ConsumerComponent cb={cbMock} config={config} />
      </RecoilRoot>
    )
    await flushPromises()
    expect(cbMock).toHaveBeenCalledTimes(3)
    const result = cbMock.mock.calls[2][0]
    expect(result.results.length).toBe(1)
    expect(result.results[0].valid).toBe(false)
    expect(result.results[0].errorMessage).toBe('Some error message')
    expect(validationDataStore.has(['page'])).toBe(true)
  })
  it.each([ null, undefined ])('should error if %o config is supplied', async (noConfig) => {
    propDataStore.set(['page'], true, 'the-page-data')
    expect(() => render(
      <RecoilRoot>
        <ConsumerComponent
          // @ts-ignore
          config={noConfig} />
      </RecoilRoot>
    )).toThrowErrorMatchingSnapshot()
  })
})
export {}