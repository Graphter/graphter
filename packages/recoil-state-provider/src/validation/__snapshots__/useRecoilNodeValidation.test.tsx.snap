// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRecoilNodeValidation should error if null config is supplied 1`] = `"Cannot read property 'validation' of null"`;

exports[`useRecoilNodeValidation should error if undefined config is supplied 1`] = `"Cannot read property 'validation' of undefined"`;

exports[`useRecoilNodeValidation should throw an error if the node has no state 1`] = `"Couldn't find state at page"`;
